pkg/reconciler/taskrun/taskrun.go

```diff
@@ -481,31 +481,35 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1beta1.TaskRun, rtr *re
 		}
 	}
 
-	if pod == nil {
-		if tr.HasVolumeClaimTemplate() {
-			if err := c.pvcHandler.CreatePersistentVolumeClaimsForWorkspaces(ctx, tr.Spec.Workspaces, *kmeta.NewControllerRef(tr), tr.Namespace); err != nil {
-				logger.Errorf("Failed to create PVC for TaskRun %s: %v", tr.Name, err)
-				tr.Status.MarkResourceFailed(volumeclaim.ReasonCouldntCreateWorkspacePVC,
-					fmt.Errorf("Failed to create PVC for TaskRun %s workspaces correctly: %s",
-						fmt.Sprintf("%s/%s", tr.Namespace, tr.Name), err))
-				return controller.NewPermanentError(err)
-			}
-
-			taskRunWorkspaces := applyVolumeClaimTemplates(tr.Spec.Workspaces, *kmeta.NewControllerRef(tr))
-			// This is used by createPod below. Changes to the Spec are not updated.
-			tr.Spec.Workspaces = taskRunWorkspaces
+	if pod == nil && tr.HasVolumeClaimTemplate() {
+		if err := c.pvcHandler.CreatePersistentVolumeClaimsForWorkspaces(ctx, tr.Spec.Workspaces, *kmeta.NewControllerRef(tr), tr.Namespace); err != nil {
+			logger.Errorf("Failed to create PVC for TaskRun %s: %v", tr.Name, err)
+			tr.Status.MarkResourceFailed(volumeclaim.ReasonCouldntCreateWorkspacePVC,
+				fmt.Errorf("Failed to create PVC for TaskRun %s workspaces correctly: %s",
+					fmt.Sprintf("%s/%s", tr.Namespace, tr.Name), err))
+			return controller.NewPermanentError(err)
 		}
 
-		// Get the randomized volume names assigned to workspace bindings
-		workspaceVolumes := workspace.CreateVolumes(tr.Spec.Workspaces)
+		taskRunWorkspaces := applyVolumeClaimTemplates(tr.Spec.Workspaces, *kmeta.NewControllerRef(tr))
+		// This is used by createPod below. Changes to the Spec are not updated.
+		tr.Spec.Workspaces = taskRunWorkspaces
+	}
 
-		ts, err := applyParamsContextsResultsAndWorkspaces(ctx, tr, rtr, workspaceVolumes)
-		if err != nil {
-			logger.Errorf("Error updating task spec parameters, contexts, results and workspaces: %s", err)
-			return err
-		}
-		tr.Status.TaskSpec = ts
+	// Get the randomized volume names assigned to workspace bindings
+	workspaceVolumes := workspace.CreateVolumes(tr.Spec.Workspaces)
 
+	ts, err := applyParamsContextsResultsAndWorkspaces(ctx, tr, rtr, workspaceVolumes)
+	if err != nil {
+		logger.Errorf("Error updating task spec parameters, contexts, results and workspaces: %s", err)
+		return err
+	}
```

Comment:
Do these need to be computed each time, or does this only need to happen if pod == nil?

This might help clean things up a little bit if we can preserve the old structure so we can keep all the `if pod == nil` logic together - i.e.

```go
if pod == nil {
  if tr.HasVolumeClaimTemplate() {
    ...
  }
 
  workspaceVolumes := workspace.CreateVolumes(tr.Spec.Workspaces)
  ...

  pod, err = c.createPod(ctx, ts, tr, rtr, workspaceVolumes)
  ...
}

tr.Status.TaskSpec = ts
```

Reply:
The problem is that we need `workspaceVolumes` as input to `applyParamsContextsResultsAndWorkspaces`, and the whole issue here stemmed from that not getting called and `tr.Status.TaskSpec` not getting updated, so I think we do need to the `createVolumes` call every time. That's what was being done before https://github.com/tektoncd/pipeline/commit/9cf7590959acd9ad9518e62d56f5941885614446